src/librustdoc/html/render/mod.rs

```diff
@@ -738,41 +738,58 @@ fn render_impls(
 }
 
-fn naive_assoc_href(it: &clean::Item, link: AssocItemLink<'_>, cx: &Context<'_>) -> String {
-    use crate::formats::item_type::ItemType::*;
+/// Build a possibly empty `href` attribute (a key-value pair) for the given associated item.
+fn assoc_href_attr(it: &clean::Item, link: AssocItemLink<'_>, cx: &Context<'_>) -> String {
+    let name = it.name.unwrap();
+    let item_type = it.type_();
 
-    let name = it.name.as_ref().unwrap();
-    let ty = match it.type_() {
-        Typedef | AssocType => AssocType,
-        s => s,
-    };
+    let href = match link {
+        AssocItemLink::Anchor(Some(ref id)) => Some(format!("#{}", id)),
+        AssocItemLink::Anchor(None) => Some(format!("#{}.{}", item_type, name)),
+        AssocItemLink::GotoSource(did, provided_items) => {
+            use ItemType::*;
+
+            let (ty_provided_item, ty_required_item, kind) = match item_type {
+                Method | TyMethod => (Method, TyMethod, ty::AssocKind::Fn),
+                AssocConst | TyAssocConst => (AssocConst, TyAssocConst, ty::AssocKind::Const),
+                AssocType | TyAssocType => (AssocType, TyAssocType, ty::AssocKind::Type),
+                _ => panic!("assoc_href_attr called on non-associated item"),
+            };
 
-    let anchor = format!("#{}.{}", ty, name);
-    match link {
-        AssocItemLink::Anchor(Some(ref id)) => format!("#{}", id),
-        AssocItemLink::Anchor(None) => anchor,
-        AssocItemLink::GotoSource(did, _) => {
-            href(did.expect_def_id(), cx).map(|p| format!("{}{}", p.0, anchor)).unwrap_or(anchor)
+            // We're creating a link from an impl-item to the corresponding
+            // trait-item and need to map the anchored type accordingly.
+            let provided = provided_items.contains(&(name, kind));
+            let ty = if provided { ty_provided_item } else { ty_required_item };
+
+            match href(did.expect_def_id(), cx) {
+                Ok((url, ..)) => Some(format!("{}#{}.{}", url, ty, name)),
+                Err(HrefError::DocumentationNotBuilt) if ty == ty_required_item => None,
+                Err(_) => Some(format!("#{}.{}", ty, name)),
+            }
         }
-    }
+    };
+
+    // links without a href are valid:
+    // https://html.spec.whatwg.org/multipage/links.html#links-created-by-a-and-area-elements
+    href.map(|href| format!(" href=\"{}\"", href)).unwrap_or_default()
 }
 
 fn assoc_const(
     w: &mut Buffer,
     it: &clean::Item,
     ty: &clean::Type,
     link: AssocItemLink<'_>,
-    extra: &str,
+    indent: usize,
     cx: &Context<'_>,
 ) {
     write!(
         w,
-        "{}{}const <a href=\"{}\" class=\"constant\">{}</a>: {}",
-        extra,
-        it.visibility.print_with_space(it.def_id, cx),
-        naive_assoc_href(it, link, cx),
-        it.name.as_ref().unwrap(),
-        ty.print(cx)
+        "{indent}{vis}const <a{href} class=\"constant\">{name}</a>: {ty}",
+        indent = " ".repeat(indent),
```

Comment:
I tihnk it'd be much better to keep the "old" way so that it's not performed for every associated constant we display.